neighsyncd/neighsync.cpp

```diff
@@ -11,12 +11,244 @@
+        }
+    }
+}
+
+NeighRestartAssit::~NeighRestartAssit()
+{
+}
+
+/* join the field-value strings for straight printing */
+string NeighRestartAssit::joinVectorString(const vector<FieldValueTuple> &fv)
+{
+    string s;
+    for (const auto &temps : fv )
+    {
+	   s += temps.first + ":" + temps.second + ", ";
+    }
+    return s;
+}
+
+void NeighRestartAssit::setCacheEntryState(std::vector<FieldValueTuple> &fvVector,
+    cache_state_t state)
+{
+    fvVector.back().second = cacheStateMap.at(state);
+}
+
+NeighRestartAssit::cache_state_t NeighRestartAssit::getCacheEntryState(const std::vector<FieldValueTuple> &fvVector)
+{
+    for (auto &iter : cacheStateMap)
+    {
+        if (fvVector.back().second == iter.second)
+        {
+            return iter.first;
+        }
+    }
+    assert("Cache Entry State is corrupted" == NULL);
+    return UNKNOWN;
+}
+
+/* Read table from APPDB and append stale flag then insert to cachemap */
+void NeighRestartAssit::readTableToMap()
+{
+    vector<string> keys;
+
+    m_appTable.getKeys(keys);
+    FieldValueTuple state(CACHE_STATE_FIELD, "");
+
+    for (const auto &key: keys)
+    {
+        vector<FieldValueTuple> fv;
+
+	    /* if the fieldvlaue empty, skip */
+        if (!m_appTable.get(key, fv))
+        {
+            continue;
+        }
+
+        fv.push_back(state);
+        setCacheEntryState(fv, STALE);
+
+        string s = "";
+        s = joinVectorString(fv);
+
+        SWSS_LOG_INFO("write to cachemap: %s, key: %s, "
+               "%s", m_appTableName.c_str(), key.c_str(), s.c_str());
+
+        // insert to the cache map
+        neighborCacheMap[key] = fv;
+    }
+    SWSS_LOG_NOTICE("Restored appDB table to internal cache map");
+    return;
+}
+
+void NeighRestartAssit::insertToMap(string key, vector<FieldValueTuple> fvVector, bool delete_key)
+{
+    string s;
+    s = joinVectorString(fvVector);
+
+    SWSS_LOG_INFO("Received message %s, key: %s, "
+            "%s, delete = %d", m_appTableName.c_str(), key.c_str(), s.c_str(), delete_key);
+
+    /*
+     * Check and insert to CacheMap Logic:
+     * if delete_key, mark the entry as "DELETE";
+     * else:
+     *  if key exist {
+     *    if it has different value: update with "NEW" flag.
+     *    if same value:  mark it as "SAME";
+     *  } else {
+     *    insert with "NEW" flag.
+     *   }
+     */
+
+    NeighborTableMap::iterator found = neighborCacheMap.find(key);
+
+    if (delete_key)
+    {
+        SWSS_LOG_NOTICE("%s, delete key: %s, ", m_appTableName.c_str(), key.c_str());
+        /* mark it as DELETE if exist, otherwise, no-op */
+        if (found != neighborCacheMap.end())
+        {
+            setCacheEntryState(found->second, DELETE);
+        }
+    }
+    else if (found != neighborCacheMap.end())
+    {
+        /* check only the original vector range (exclude cache-state field/value) */
+        if(!equal(fvVector.begin(), fvVector.end(), found->second.begin()))
+        {
```

Comment:
Is it possible that  the order of field/value has been changed though the internal content is the same?  ex.  f1/v1,f2/v2,  got changed to f2/v2, f1/v1.   equal algorithm will take them as different fvvector.